imagecraft/pack/chroot.py

```diff
@@ -0,0 +1,189 @@
+# Copyright 2025 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Execute a callable in a chroot environment."""
+
+import logging
+import multiprocessing
+import os
+from collections.abc import Callable
+from multiprocessing.connection import Connection
+from pathlib import Path
+from typing import Any
+
+from craft_cli import emit
+from craft_parts.utils import os_utils
+
+from imagecraft import errors
+
+logger = logging.getLogger(__name__)
+
+
+class Mount:
+    """Mount entry for chroot setup."""
+
+    _fstype: str | None
+    _src: str
+    _relative_mountpoint: str
+    _options: list[str] | None = None
+    _mountpoint: Path | None = None
+
+    def __init__(
+        self,
+        fstype: str | None,
+        src: str,
+        relative_mountpoint: str,
+        *,
+        options: list[str] | None = None,
+        mountpoint: Path | None = None,
+    ) -> None:
+        self._fstype = fstype
+        self._src = src
+        self._relative_mountpoint = relative_mountpoint
+        if options:
+            self._options = options
+        if mountpoint:
+            self._mountpoint = mountpoint
+
+    def mount(self, base_path: Path) -> None:
+        """Mount the mountpoint.
+
+        :param base_path: path to mount the mountpoint under.
+        """
+        args: list[str] = []
+        if self._options:
+            args.extend(self._options)
+        if self._fstype:
+            args.append(f"-t{self._fstype}")
+
+        self._mountpoint = base_path / self._relative_mountpoint.lstrip("/")
+        pid = os.getpid()
+        # Only mount if mountpoint exists.
+        emit.debug(f"Mount {self._src} on {str(self._mountpoint)}")
+        logger.debug("[pid=%d] mount %r on chroot", pid, str(self._mountpoint))
+        if self._mountpoint.exists():
+            logger.debug("[pid=%d] mount %r on chroot", pid, str(self._mountpoint))
+            os_utils.mount(self._src, str(self._mountpoint), *args)
+        else:
```

Comment:
If someone asks for a mount operation, and its not done here, will that not cause a failure further down the line? Is this not an error condition that is better caught early ?

Reply:
Good point. And also the umounting should probably not stop at the first umounting failure but try to umount as many mountpoints as possible before raising an exception if at least one could not be umounted. I created an [issue](https://github.com/canonical/imagecraft/issues/118) for that to deal with that in another PR.